python/distributed-ucxx/distributed_ucxx/ucxx.py

```diff
@@ -90,13 +90,84 @@ def synchronize_stream(stream=0):
     stream.synchronize()
 
 
+def _stop_notifier_thread_and_progress_tasks():
+    """Stop the notifier thread and progress tasks.
+
+    If no Dask resources that make use of UCXX communicator are running anymore,
+    stop the notifier thread and progress tasks to allow for clean shutdown.
+    """
+    ctx = ucxx.core._get_ctx()
+    if len(ctx._dask_resources) == 0:
+        ctx.stop_notifier_thread()
+        ctx.progress_tasks.clear()
+
+
+def _register_dask_resource(resource):
+    """Register a Dask resource with the UCXX context.
+
+    Register a Dask resource with the UCXX context to keep track of it, so that
+    the notifier thread and progress tasks may be stopped when no more resources
+    """
+    ctx = ucxx.core._get_ctx()
+
+    with ctx._dask_resources_lock:
+        ctx._dask_resources.add(resource)
+        ctx.start_notifier_thread()
+        ctx.continuous_ucx_progress()
+
+
+def _deregister_dask_resource(resource):
+    """Deregister a Dask resource with the UCXX context.
+
+    Deregister a Dask resource from the UCXX context, and if no resources remain
+    after deregistration, stop the notifier thread and progress tasks.
+    need UCXX.
+    """
+    if ucxx.core._ctx is None:
+        # Prevent creation of context if it was already destroyed, all
+        # registered references are already gone.
+        return
+
+    ctx = ucxx.core._get_ctx()
+
+    # Check if the attribute exists first, in tests the UCXX context may have
+    # been reset before some resources are deregistered.
+    if hasattr(ctx, "_dask_resources_lock"):
+        with ctx._dask_resources_lock:
+            try:
+                ctx._dask_resources.remove(resource)
+            except KeyError:
+                pass
+            _stop_notifier_thread_and_progress_tasks()
+
+
+def _allocate_dask_resources_tracker() -> None:
+    """Allocate Dask resources tracker.
+
+    Allocate a Dask resources tracker in the UCXX context. This is useful to
+    track Distributed communicators so that progress and notifier threads can
+    be cleanly stopped when no UCXX communicators are alive anymore.
+    """
+    ctx = ucxx.core._get_ctx()
+    if not hasattr(ctx, "_dask_resources"):
+        from threading import Lock
+
+        ctx._dask_resources = set()
+        ctx._dask_resources_lock = Lock()
```

Comment:
*note*: If multiple threads are running there is a potential for a race condition here, but I'm not sure if it matters.

Reply:
Specifically: suppose we don't have the attribute and two threads are running.

Both threads can enter line 172. Then suppose thread A continues, and sets the attributes, and continues using the lock to update `_dask_resources`. Eventually thread B goes forward to line 174, it now overwrites `_dask_resources`, so we could lose any resource tags that thread A had added.

Reply:
Good point, you're indeed correct. Initially I only had `_allocate_dask_resources_tracker()` being called after `ucxx.init(...)`, but later I found the need to add it also to https://github.com/rapidsai/ucxx/pull/291/files#diff-e8b6ecb786c905744fe6e10d9547e9aaffbd732fb70750d5385ac9641babb924R185-R189 due to tests resetting the context.

Currently, I don't think this can happen in a different thread though but I may be overlooking something. One way I think we could prevent risking concurrency on creating the lock is to assign the lock and resource set to this file, such that it is only created once next to `ucxx.init(...)` instead of pushing it into the `ApplicationContext`, technically even that isn't thread-safe but I don't think we ever had a problem with it so it's "safe". If you think it's worth it I can do that but would prefer to address it in a follow-up PR as it will definitely need me to do some more robust testing again on that and I'd still like to get this integrated soon as it's already better than it was before.

With the above being said, I realize now there's a different issue here which is the impossibility of starting a new test that uses a different set of UCX variables in the same process, because `ucxx.init()` is still only called once. I think this is currently not a problem though, as the only such tests are spawned in new processes in https://github.com/rapidsai/ucxx/blob/branch-0.41/python/distributed-ucxx/distributed_ucxx/tests/test_ucxx_config.py .

Reply:
> on creating the lock is to assign the lock and resource set to this file, such that it is only created once next to ucxx.init(...) instead of pushing it into the ApplicationContext

You mean, make the lock a file-level (i.e. module-level variable)? That is thread safe because `import` is protected by a lock in the python interpreter.

Overall, I think as is, this is fine. Perhaps add a comment that creating the comm objects is not thread safe, so we remember if in the future we do need thread safety.

Reply:
Added a comment in https://github.com/rapidsai/ucxx/pull/291/commits/48fa0e5ebc39de249434877db6db3440fb842f33 . I'll nevertheless try to address this in a follow-up PR today or next week.